### ISSUE TITLE
VERCEL env causes signinUrl and others to be prefixed with https in local development

### ISSUE THREAD

**Issue by dferber90** [nextauthjs/next-auth#7202]
### Environment

  System:
    OS: macOS 13.3
    CPU: (10) arm64 Apple M1 Pro
    Memory: 5.52 GB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 17.4.0 - ~/Library/Caches/fnm_multishells/16656_1681032588306/bin/node
    Yarn: 1.22.15 - ~/Library/Caches/fnm_multishells/16656_1681032588306/bin/yarn
    npm: 8.3.1 - ~/Library/Caches/fnm_multishells/16656_1681032588306/bin/npm
  Browsers:
    Chrome: 112.0.5615.49
    Firefox: 111.0.1
    Safari: 16.4

### Reproduction URL

can be reproduced with next-auth-example by adding VERCEL="1" env var

### Describe the issue

next-auth fails if an `VERCEL="1"` env var is present - even if it's set to `VERCEL="0"`.

I use `vc env pull .env.local` to pull the Vercel defined env vars to my machine.

This creates a `VERCEL="1"` env var. But as soon as this is present the sign in no longer works as
I get redirected to `https://localhost:3000/api/auth/signin?csrf=true` (notice the `https`) when trying to sign in.

I debugged this and noticed the following: If you open `http://localhost:3000/api/auth/providers` you will see this

```json
{
  "github": {
    "id": "github",
    "name": "GitHub",
    "type": "oauth",
    "signinUrl": "https://localhost:3000/api/auth/signin/github",
    "callbackUrl": "https://localhost:3000/api/auth/callback/github"
  }
}
```

Note that `signinUrl` and `callbackUrl` are prefixed with `https`. This prefix goes away as soon as you remove the `VERCEL="1"` env var.

### How to reproduce

Create a `VERCEL="1"` env var and try to sign in.

Next.js version 13.3.0
next-auth version 4.21.1

### Expected behavior

Sign in should work. It should not redirect to https locally.

_This slightly relates to https://github.com/nextauthjs/next-auth/issues/4363_

**Comment by Crazypersonalph** [comment]
I have had this issue before.
This is because the cookies you are creating for authentication/session stuff are HTTPS only, so they are not allowed to be used with HTTP only, therefore causing the redirect.

**Comment by Crazypersonalph** [comment]
I know in SvelteKitAuth, which is just this project being wrapped for sveltekit, there is an option called ```useSecureCookies```, by default it is on in production, to protect the cookie information, but you can manually disable it.

**Comment by Crazypersonalph** [comment]
It may also be that Vercel is changing around the links by itself without your knowledge, keep an eye out for that.

**Comment by dferber90** [comment]
I don't think it's related to `useSecureCookies` as it can be reproduced in an incognito window. This is specifically about the `VERCEL` env var being present.

**Comment by Crazypersonalph** [comment]
Looks like this may hold an answer:
https://github.com/vercel/vercel/issues/1745

**Comment by dferber90** [comment]
It's not related. This is really about the use of `process.env.VERCEL` within the next-auth code base itself. It is even reproducible locally, so it's not an issue of being hosted on Vercel <picture data-single-emoji=":blob_smile:" title=":blob_smile:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..2yi5QO7-oXo7z4xN.TP3deKYXCw6FwTnN8aJM2cbY1h8aKNAa5PNVEniTqqkWncfaMa4K9EgGNmQfqvjoQ3QDgoqOlOcsQjf7FTAptjqwoh3TJnGNMTIrIRyoW9cKZ2mZt_SNL9jM_50.Muxv3gFIHSKCxTYOb1RIwg" alt=":blob_smile:" align="absmiddle"></picture>

**Comment by Crazypersonalph** [comment]
> It's not related. This is really about the use of `process.env.VERCEL` within the next-auth code base itself. It is even reproducible locally, so it's not an issue of being hosted on Vercel <img alt=":blob_smile:" width="20" height="auto" src="https://camo.githubusercontent.com/a96ea7179de8ce4bd971b96128d30f0a7e89e83ebfee239de112952432a84e74/68747470733a2f2f73696e676c652d656d6f6a692e76657263656c2e6170702f6170692f656d6f6a692f65794a68624763694f694a6b615849694c434a6c626d4d694f694a424d6a553252304e4e496e302e2e32796935514f372d6f586f377a34784e2e54503364654b59584377364677546e4e38614a4d32636259316838614b4e416135504e56456e695471716b576e6366614d61344b394567474e6d516671766a6f51335144676f714f6c4f6373516a663746544170746a71776f6833544a6e474e4d5449724952796f5739634b5a326d5a745f534e4c396a4d5f35302e4d7578763367464948534b437854594f623152497767">
> 
> _Sorry if it wasn't clear earlier: I work at Vercel and I'll get this sorted internally too :)_

I know this isn't related, but that is really cool! I am a student just playing around in the web dev space, and normally I use C++. Thank you for your contribution!

**Comment by fsa317** [comment]
I am having the same issue @dferber90 .  Is there a workaround?

**Comment by fsa317** [comment]
From what I can see it looks like this code uses that env variable to determine if to use secure cookies:

https://github.com/nextauthjs/next-auth/blob/62f672ae30fb389c1a81f484f2dad7a3e2a6a662/packages/next-auth/src/jwt/index.ts#L73

**Comment by Crazypersonalph** [comment]
> From what I can see it looks like this code uses that env variable to determine if to use secure cookies:
> 
> https://github.com/nextauthjs/next-auth/blob/62f672ae30fb389c1a81f484f2dad7a3e2a6a662/packages/next-auth/src/jwt/index.ts#L73

That is what I was thinking about before

**Comment by dferber90** [comment]
> I am having the same issue @dferber90 . Is there a workaround?

The workaround I used was to comment out the `VERCEL` env var in my `.env.local` file manually. It's not a great workaround but at least it unblocked me.

@Crazypersonalph I was trying to say that it's not related to any cookies I already have, but the error is definitely somewhere in the logic of how next-auth detects whether to use https or not.

**Comment by balazsorban44** [comment]
I found the issue, and created an experimental release, if anyone wants to test https://github.com/nextauthjs/next-auth/pull/7214#issuecomment-1503033702

TLDR; we need to respect the `x-forwarded-proto` header if present, when trusting the host. (In this case when `VERCEL` environment variable is set)

**Comment by Crazypersonalph** [comment]
It is safe to say that this issue can now be closed.

**Comment by balazsorban44** [comment]
This should be released in `4.22.1`!

**Comment by MarcinCebula** [comment]
Hey @balazsorban44 

There is a case where respecting x-forwarded-proto it not true. When using Zones the url returned would be wrong and would cause multiple issue.

1. you would get the wrong domian
2. you would get the wrong basePath

For example 
https://github.com/nextauthjs/next-auth/blob/c7dec376a1272a0877c7e6fb2fc4173f08369f9f/packages/next-auth/src/utils/detect-origin.ts#LL5C1-L6C1

The desired url might be https://mydomain/app2/auth/.... 
btw when doing SignIn from path 2 using Zones on Vercel you would get https://app2.mydomain/auth/...

Please notice the subdomain and basePath 

NOTES:
1. Things to note. For zones to work correctly most people extract x-forwarded-proto in middleware and set it to some other header like x-app-root-url
2. A hackey solution to quickly solve this issue is to set VERCEL="" in ENV


Would love to discuss a solution for this issue and maybe create a example project to showcase the issue

**Comment by montnyc** [comment]
> > I am having the same issue @dferber90 . Is there a workaround?
> 
> The workaround I used was to comment out the `VERCEL` env var in my `.env.local` file manually. It's not a great workaround but at least it unblocked me.
> 
> @Crazypersonalph I was trying to say that it's not related to any cookies I already have, but the error is definitely somewhere in the logic of how next-auth detects whether to use https or not.

This is the only thing that worked for me... Was banging my head against the wall. Thanks! Would love an actual fix tho

**Comment by leogilmour** [comment]
We are still having this issue. Finally figured it out and commented `VERCEL="1"` out
```"next": "^13.4.19",
    "next-auth": "^4.23.1",```

**Comment by toniopelo** [comment]
@dferber90 @balazsorban44 @leogilmour There is a more convenient workaround I just tested, to be able to still use `vercel env pull` without having to comment out `VERCEL="1"` every time. Just add an env var for development only on vercel platform, name this env var `VERCEL` and leave the value blank. This is pretty dumb but it works without breaking vercel env updating flow.

![image](https://github.com/nextauthjs/next-auth/assets/9287511/d1c4b25a-7035-42c1-bae6-6699c6cb1f29)

A fix on nextauth side is still needed IMO because this shouldn't be needed.

**Comment by fsa317** [comment]
I seem to still be having this issue on the latest NEXT_AUTH

**Comment by dipsor** [comment]
I am having same issue

**Comment by gloriliale** [comment]
Same here!

**Comment by aichi-t** [comment]
I'm also having the same issue, even after removing `VERCEL="1"` from .env, on version `4.23.1`. 

Does anyone know a fix for this issue?

**Comment by reginpv** [comment]
> > I am having the same issue @dferber90 . Is there a workaround?
> 
> The workaround I used was to comment out the `VERCEL` env var in my `.env.local` file manually. It's not a great workaround but at least it unblocked me.
> 
> @Crazypersonalph I was trying to say that it's not related to any cookies I already have, but the error is definitely somewhere in the logic of how next-auth detects whether to use https or not.

This is what I do currently, but is there a permanent fix? If I `vercel env pull` it will change again to 1

**Comment by reginpv** [comment]
> @dferber90 @balazsorban44 @leogilmour There is a more convenient workaround I just tested, to be able to still use `vercel env pull` without having to comment out `VERCEL="1"` every time. Just add an env var for development only on vercel platform, name this env var `VERCEL` and leave the value blank. This is pretty dumb but it works without breaking vercel env updating flow.
> 
> ![image](https://private-user-images.githubusercontent.com/9287511/271839180-d1c4b25a-7035-42c1-bae6-6699c6cb1f29.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4NDYwOTIsIm5iZiI6MTcyMDg0NTc5MiwicGF0aCI6Ii85Mjg3NTExLzI3MTgzOTE4MC1kMWM0YjI1YS03MDM1LTQyYzEtYmFlNi02Njk5YzZjYjFmMjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTNUMDQ0MzEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTg5ZTUwYzM3YzI1YjhmZGQxYjVhMWIxMTFlNzkwZjAwNzU3MzE5NTYxNjJlNzBkNTVjZWY4OTQwODRmMmJhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.B_xkToSP0il5fInwOmshfmYzL7Bq_fDGZfR4pbHps3c)
> 
> A fix on nextauth side is still needed IMO because this shouldn't be needed.

This is the one that works for me! Thank you!
